docs/incubator-proposal.md

```diff
@@ -0,0 +1,194 @@
+# cert-manager proposal
+
+## Problem
+
+Currently in Kubernetes, a number of different components including user
+applications require TLS certificates. Right now, users can provide
+certificates to workloads on the cluster by creating `Secret` resources that
+contain a private/public keypair.
```

Comment:
This isn't the only way though, and it's must less secure than generating a CSR on workload instance boot. This model also doesn't allow hardware enforced non-exporability of keys.

Reply:
See https://github.com/kelseyhightower/certificate-init-container/issues/3#issuecomment-318715289

Reply:
Ack.

Does storing a signed certificate in a secret still make sense here? Otherwise we may need to consider making the delivery mechanism for signed certificates generic.

Supporting a mode of operation where the private key can be kept secret, and instead a public key is provided to be used with the CSR, or a CSR itself provided would be great. We could follow the `certificates.k8s.io` API design and include the signed Certificate in the status block of the resource potentially too, and then optionally export that to a secret if requested.

In this mode of operation, how would renewals work? Right now the CSR itself is generated by cert-manager, so it is able to set the date fields appropriately. This makes me think we want to expose a `csrTemplate` of some sort in our spec perhaps.

Reply:
I do like the idea of allowing users to provide their own CSRs - perhaps the `certificates.k8s.io` API group can be used as an underlying basis for cert-manager?

If we add a `authority`/`issuer`/`signer` field to the CertificateSigningRequest resource, cert-manager could then issue these certificates then cert-manager would be able to selectively choose how to issue each certificate.

However, this does not solve the issue of providing certificate-specific configuration for each CSR (e.g. which challenge mechanism to use for each domain in the ACME case)